packages/hardhat-core/src/internal/core/runtime-environment.ts

```diff
@@ -260,23 +269,39 @@ export class Environment implements HardhatRuntimeEnvironment {
       runSuperFunction.isDefined = false;
     }
 
-    const runSuper: RunSuperFunction<TaskArguments> = runSuperFunction;
+    const runSuper: RunSuperFunction<TaskArguments, SubtaskArguments> =
+      runSuperFunction;
 
     const globalAsAny = global as any;
     const previousRunSuper: any = globalAsAny.runSuper;
     globalAsAny.runSuper = runSuper;
 
-    let modifiedHreWithParentTaskProfile: any | undefined;
+    // We create a modified version of `this`, as we want to keep track of the
+    // `subtaskArguments` and use it to generate updated `subtaskArguments` if
+    // the action calls `run`.
+    //
+    // Note that for this to work we need to set the prototype later
+    let modifiedHreWithParentTaskProfile: any = {
```

Comment:
```suggestion
    let modifiedHre: any = {
```

This name is outdated.